### PR TITLE
 FP8 Megablox for batch split

### PR DESCRIPTION
**Pull request by BirdsOfAFthr**
# Description

### (1) This update enables FP8 Megablox quantization support for DeepSeek batch split configurations.

When quantization is active, the following changes apply:

* Kernel Quantization: gmm kernels allow FP8 recipes (defined via the MaxText command line) in both forward and backward passes.

* `gmm` forward: weight is manually quantized to bypass the expcilt sharding error; activation is quantized using qwix
* `gmm` backward: gradients are quantized using qwix.
### (2) This change also enables merging gating gmm kernels.
In the previous SwiGLU/GLU implementation, the gate-projection and up-projection were processed using two sequential `gmm_fn` calls. By concatenating these weights and processing them together, we effectively double the contiguous hidden dimension of the kernel. This is especially critical for FP8 utilizing Expert Parallelism (EP) that shard along the contracting dimension. Because this sharding strategy inherently shrinks the local MLP hidden dimension on each device, the matrix multiplications can become small and bottlenecked by memory bandwidth. Merging $W_0$ and $W_1$ effectively gives us a 2X increase in that local dimension, restoring arithmetic intensity and hardware utilization.

# Tests

* Verification: Validated via end-to-end (e2e) perf and convergence benchmarks.

* Coverage: Unit tests will be added in a subsequent update.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by gobbleturk** [comment]
Note there is more general support in this PR https://github.com/AI-Hypercomputer/maxtext/pull/3736

File: src/maxtext/kernels/megablox/ops.py [hunk 1]
```diff
@@ -169,6 +176,9 @@ def _gmm_fwd(
         preferred_element_type=preferred_element_type,
         group_offset=group_offset,
         implementation="mosaic",
+        manual_axis_type=jax.sharding.ManualAxisType(varying=frozenset(["data", "fsdp"]))
```
Comment (shuningjin) [inline src/maxtext/kernels/megablox/ops.py]:
```
manual_axis_type=jax.sharding.ManualAxisType(varying=frozenset(["data", "fsdp", "expert"]))
```

File: src/maxtext/kernels/megablox/ops.py [hunk 2]
```diff
@@ -267,6 +281,9 @@ def _gmm_bwd(
         preferred_element_type=rhs_dtype,
         group_offset=group_offset,
         implementation="mosaic",
+        manual_axis_type=jax.sharding.ManualAxisType(varying=frozenset(["data", "fsdp", "expert"]))
```
Comment (shuningjin) [inline src/maxtext/kernels/megablox/ops.py]:
```
manual_axis_type=jax.sharding.ManualAxisType(varying=frozenset(["expert"]), unreduced=frozenset(["data", "fsdp"]))
```

File: src/maxtext/layers/quantizations.py [hunk 3]
```diff
@@ -851,6 +849,72 @@ def maybe_quantize_model(model, config):
...
+
+
+def _get_max_min(target_dtype):
+  if target_dtype in (jnp.int4, jnp.int8):
+    return jnp.iinfo(target_dtype).max, jnp.iinfo(target_dtype).min
+  else:
+    return jnp.finfo(target_dtype).max.astype(jnp.bfloat16), jnp.finfo(target_dtype).min.astype(jnp.bfloat16)
+
+
+def manual_quantize(tensor, calibration_method):
```
Comment (shuningjin) [inline src/maxtext/layers/quantizations.py]: nit: maybe make "fwd_dtype" an input argument for flexibility (instead of hardcoding fwd_dtype = jnp.float8_e4m3fn below), and to be consistent with how we use "qpl.quantize"

File: src/maxtext/models/deepseek_batchsplit.py [hunk 4]
```diff
@@ -166,9 +167,16 @@ def fn(weights):
...
+
+    # TODO(jesselu): Implement fp8 gmm weight AGs here.
+    # similar logic to kernels.megablox.op._gmm_fwd QAG
+    # raise NotImplementedError("fp8 gmm weight AGs are not implemented yet.")
+
+    # Quantized weights are PyTrees (f8 data + scales)?
+    routed_wi_0 = jax.tree.map(lambda x: jax.lax.all_gather(x, axis_name="fsdp", tiled=True, to="reduced"), routed_wi_0)
+    routed_wi_1 = jax.tree.map(lambda x: jax.lax.all_gather(x, axis_name="fsdp", tiled=True, to="reduced"), routed_wi_1)
+    routed_wo = jax.tree.map(lambda x: jax.lax.all_gather(x, axis_name="fsdp", tiled=True, to="reduced"), routed_wo)
+
```
Comment (shuningjin) [inline src/maxtext/models/deepseek_batchsplit.py]: This change is not needed for our PR; could revert.

**Review by shuningjin — COMMENTED** [review]
Thanks for GMM FP8 support with careful manual quantization, while bringing back merging gating from [PR#3199](https://github.com/AI-Hypercomputer/maxtext/pull/3199)! Had a comment wrt sharding, and other minor changes.

File: src/maxtext/kernels/megablox/ops.py (hunk 1 above)
Comment (BirdsOfAFthr) [reply src/maxtext/kernels/megablox/ops.py]: good eye. It's odd that this also passed unit test

File: src/maxtext/models/deepseek_batchsplit.py [hunk 5]
```diff
@@ -1879,35 +1835,83 @@ def unroute_impl_bwd(expert_axis_name, use_gather_mosaic_kernel, res, grad):
...
+    kernel: The right-hand side operand (kernel) of the GMM.
+    group_sizes: An array indicating the size of each group.
+    preferred_element_type: The preferred element type for the computation.
+    config: Configuration object containing model settings, including
+      `use_qwix_quantization` and `merge_gating_gmm`.
+
+  Returns:
+    The result of the grouped matrix multiplication.
+  """
+  if config.use_qwix_quantization:
```
Comment (shuningjin) [inline src/maxtext/models/deepseek_batchsplit.py]: maybe this condition is clearer?
```
if config.quantization == "fp8_full":
```

Comment (BirdsOfAFthr) [reply src/maxtext/models/deepseek_batchsplit.py]: Using `use_qwix_quantization` to check quant is the same as using `config.quantization`, but we don't check if it is `fp8_full`, as we can use other recipes as well.

`use_qwix_quantization` is a flag that was introduced during AQT->Qwix migration.

**Review by shuningjin — APPROVED** [review]

**Review by suexu1025 — APPROVED** [review]
Thanks @BirdsOfAFthr
